docs/proposals/project-repos-and-clusters.md

```diff
@@ -0,0 +1,154 @@
+and clusters are stored as Kubernetes Secrets, so a new field could be stored as a Secret data key:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: argocd-example-apps
+  labels:
+    argocd.argoproj.io/secret-type: repository
+type: Opaque
+stringData:
+  project: my-project1                                     # new project field
+  name: argocd-example-apps
+  url: https://github.com/argoproj/argocd-example-apps.git
+  username: ****
+  passoword: ****
+```
+
+* The project scoped repository/cluster is automatically allowed in the project.
+  This enables developers to allow new cluster/repository without modifying the project.
+* The project scoped repository/cluster still can be used in other project but it has to be allowed by admin (as normal repository/cluster).
+* If another team wants to add the same repository/cluster into a different project they would have to ask admin. 
+
+#### Project RBAC Changes
+
+The organization still might want to enforce certain rules so developers won't get permission to add a
+project-specific repository/cluster by default. The administrator might use RBAC to control access to
+the project scoped repositories cluster. The access to project scope actions will be checked using
+`<projectName>/<name>` pattern. For example, to allow users to add project scoped repositories admin would have to add
+the following RBAC rules:
+
+```
+p, proj:my-project:admin, repositories, create, my-project/*, allow
+p, proj:my-project:admin, repositories, delete, my-project/*, allow
+p, proj:my-project:admin, repositories, update, my-project/*, allow
+```
+
+This provides extra flexibility so that admin can have stricter rules. e.g.:
+
+```
+p, proj:my-project:admin, repositories, update, my-project/"https://github.my-company.com/*", allow
+```
+
+#### UI/CLI Changes
+
+Both User interface and CLI should get ability to optionally specify project. If project is specified than cluster/repository
+is considered project scoped:
+
+```
+argocd repo add --name stable https://charts.helm.sh/stable --type helm --project my-project
+```
+
+
+### Use cases
+
+Add a list of detailed use cases this enhancement intends to take care of.
+
+## Use case 1:
+As a developer, I would like to register credentials of a Git repository I own so I can deploy manifests stored in that repository.
+
+## Use case 2:
+As a developer, I would like to register credentials of a Kubernetes cluster so I can manage resources in that cluster.
+
+### Implementation Details/Notes/Constraints [optional]
+
+As of v2.0.1 Argo CD stores Repository non-sensitive metadata in `argocd-cm` ConfigMap. This is going to change in https://github.com/argoproj/argo-cd/issues/5436.
+So we would have to wait for #5436 implementation.
+
+### Detailed examples
+
+### Security Considerations
+
+The security considerations are explained in `Project RBAC Changes` section.
+
+### Risks and Mitigations
+
+TBD
+
```

Comment:
Self service of cluster registration opens the door for performance problems in the controller. As discussed, we'll likely need an additional project tag in prometheus so that operators can discern which projects are contributing to high usage.

Reply:
done